src/libmongoc/src/mongoc/mongoc-timeout.c

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright 2020 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mongoc-timeout-private.h"
+#include "mongoc.h"
+
+typedef struct _mongoc_timeout_t {
+   bool is_set;
+   int64_t timeout_ms;
+} mongoc_timeout_t;
+
+int64_t
+mongoc_timeout_get_timeout_ms (const mongoc_timeout_t *timeout)
+{
+   BSON_ASSERT (timeout);
+
+   return timeout->timeout_ms;
+}
+
+bool
+_mongoc_timeout_set_timeout_ms (mongoc_timeout_t *timeout, int64_t timeout_ms)
+{
+   BSON_ASSERT (timeout);
+
+   if (timeout_ms < 0) {
+      MONGOC_ERROR ("timeout must not be negative");
+      return false;
+   }
+
+   timeout->timeout_ms = timeout_ms;
+   timeout->is_set = true;
+   return true;
+}
+
+bool
+mongoc_timeout_set_timeout_ms (mongoc_timeout_t *timeout, int64_t timeout_ms)
+{
+   return _mongoc_timeout_set_timeout_ms (timeout, timeout_ms);
+}
+
+mongoc_timeout_t *
+mongoc_timeout_new ()
+{
+   mongoc_timeout_t *timeout;
+
+   timeout = (mongoc_timeout_t *) bson_malloc0 (sizeof *timeout);
+   timeout->timeout_ms = 0;
+   timeout->is_set = false;
+
+   return timeout;
+}
+
+mongoc_timeout_t *
+mongoc_timeout_new_timeout_int64 (int64_t timeout_ms)
+{
+   mongoc_timeout_t *timeout = mongoc_timeout_new ();
+
+   if (_mongoc_timeout_set_timeout_ms (timeout, timeout_ms))
+      return timeout;
+
+   mongoc_timeout_destroy (timeout);
+   return NULL;
+}
+
+mongoc_timeout_t *
+mongoc_timeout_copy (const mongoc_timeout_t *timeout)
+{
+   mongoc_timeout_t *copy = NULL;
+
+   BSON_ASSERT (timeout);
+
+   copy = mongoc_timeout_new ();
+   copy->timeout_ms = timeout->timeout_ms;
+   copy->is_set = timeout->is_set;
+
+   return copy;
+}
+
+void
+mongoc_timeout_destroy (mongoc_timeout_t *timeout)
+{
+   bson_free (timeout);
```

Comment:
I suggest making this a no-op if `timeout` is NULL. That enables the "done" pattern:
```
mongoc_timeout_t *timeout = NULL;

if (some condition) {
   goto done;
}

mongoc_timeout_destroy (timeout); /* works whether NULL or set. */
````

Reply:
I thought `bson_free` was a no-op if its argument is `NULL`. I added a `NULL` test for `mongoc_timeout_destroy` just in case it fails on some architectures. I also found an example of the "done" pattern with `bson_free` in `client-side-encryption-schema-map:create_schema_file:90`. I'm not 100% sure, though. Waiting for evg tests.

Reply:
Ah that is correct, `bson_free` follows this pattern. My comment was a knee-jerk reaction since most `destroy` methods have that explicit check. If `mongoc_timeout_t` had other fields to free it would need it. Since it does not, there's nothing to do here.